tests/foreman/virtwho/utils.py

```diff
@@ -303,3 +305,54 @@ def deploy_configure_by_script(script_content, debug=False):
         )
     if debug:
         return deploy_validation()
+
+
+def restart_virt_who_service():
+    """
+    Do the following:
+    1. clean rhsm.log message, make sure there is no old message exist.
+    2. restart virt-who service via systemctl command
+    """
+    runcmd("rm -f /var/log/rhsm/rhsm.log")
+    runcmd("systemctl restart virt-who")
+
+
+def create_etc_d_file(form_data,
+                      rhsm_username='admin',
+                      rhsm_password='admin',
+                      filename='virt-who-config.conf'):
+    filepath = '/etc/virt-who.d/{}'.format(filename)
+    type = form_data['hypervisor_type']
+    owner = 'Default_Organization'
+    env = 'Library'
+    server = form_data['hypervisor_content.server']
+    cf_file = '{0}\n'.format(filepath)
+    cf_title = '[{0}]\n'.format(filename)
```

Comment:
Please use [`ConfigParser`](https://docs.python.org/3/library/configparser.html) to create configuration structure and write it down to file, instead of hand-crafting it.
This will allow you to drop this if/else below, as you will just add specific fields based on conditions, and writ everything into file at the end.

Reply:
ok, thanks, btw, I would like to know if we have a method to assert if some elements on UI are visible/clickable or not, thanks

Reply:
Thanks for your detailed explanation, in this case, different roles for the virt-who user will cause different results. Take virt-who-viewer as an example,

> Virt-who Viewer Role granting permissions to see all configurations including their configuration scripts

In other words, when the user only configured with virt-who viewer role, the user can only see the configurations, but cannot create and edit them, which means we cannot see **create config** and **edit** button on UI, I tried this(as below) to assert the user have not permission to create config:
```
        with Session(test_name, username, password) as newsession:
            newsession.virtwho_configure.read(virt_who_name)
            try:
                viewer_virt_who_name = gen_string('alpha')
                form_data['name'] = viewer_virt_who_name
                newsession.virtwho_configure.create(form_data)
            except Exception:
                pass
            else:
                raise VirtWhoError(
                    'Virt-who Viewer should not have the permission to creat config')
```
Do you have any ideas to achieve such as:
`assert not VirtwhoConfiguresView.is_displayed()`
to make this more pythonic, thanks

Reply:
I see.

Trying to perform action and catching exception looks like a reasonable approach to me.

If you can't do it this way (because it's too slow or something), then you need to add new action to `VirtwhoConfigureEntity` in Airgun. Something like `can_create()`, or `is_editable(entity)`. There, you would navigate to specific view, check availability, visibility and classes of all elements you need and return True/False value that you can use in test for assertion. That would be something similar to [`has_manifest` in subscription entity](https://github.com/SatelliteQE/airgun/blob/master/airgun/entities/subscription.py#L43).

Reply:
ok, your suggestion is so detailed and useful, and I am considering if I should add these actions. Thanks a lot.:)
Hi, @eko999, what is your opinion? just use the way of catching exception as I said [as above](https://github.com/SatelliteQE/robottelo/pull/7405#discussion_r335352108) or add new actions in Airgun. Thanks :)

Reply:
I absolutely agree with Mirek's solution.